src/core/qgstiledownloadmanager.cpp

```diff
@@ -50,21 +50,36 @@ void QgsTileDownloadManagerWorker::queueUpdated()
 
   if ( mManager->mShuttingDown )
   {
+    // here we HAVE to build up a list of replies from the queue before do anything
+    // with them. Otherwise we can hit the situation where aborting the replies
+    // triggers immediately their removal from the queue, and we'll be modifying
+    // mQueue elsewhere while still trying to iterate over it here => crash
+    QList< QNetworkReply * > replies;
+    replies.reserve( mManager->mQueue.size() );
     for ( auto it = mManager->mQueue.begin(); it != mManager->mQueue.end(); ++it )
     {
-      it->networkReply->abort();
+      replies << it->networkReply;
+    }
+    // now abort all replies
+    for ( QNetworkReply *reply : replies )
+    {
+      reply->abort();
```

Comment:
Could we just post all those aborts into the event queue and add a mutex to prevent concurrency problems?

Reply:
It's not so simple unfortunately -- there's already a recursive mutex protecting things here, but it needs to get acquired for the queue removal. So if we make it non recursive and lock it here, we'll get a deadlock when the removal tries to acquire it. A second mutex won't help for the same reason -- we need it acquired both before the iteration and then potentially in the middle of iteration.

Reply:
If we post the removal to the event queue (like deleteLater etc) that should avoid this, no?

Reply:
Isn't that effectively the same thing, but adding a bunch of overhead?

Reply:
That would avoid quite a lot of code in here (the staging and copying the list) so I'd say it reduces maintenance overhead?

Reply:
> That would avoid quite a lot of code in here (the staging and copying the list) so I'd say it reduces maintenance overhead? But it's also possible I miss part of the picture.

See my reply to @wonder-sk below -- there's times when the event loop will be triggered while we're iterating here, so any kind of qt based deferred delete (eg deleteLater) may still end up being called out of turn. That's why I'd like to just stay in control of everything and be guaranteed that everything is getting done in the right order here, regardless of what happens in the Qt realm...